WordPress/Sniffs/WP/PostTypeSniff.php

```diff
@@ -0,0 +1,117 @@
+ */
+
+namespace WordPressCS\WordPress\Sniffs\WP;
+
+use WordPressCS\WordPress\AbstractFunctionParameterSniff;
+
+/**
+ * Validates post type names.
+ *
+ * @package WPCS\WordPressCodingStandards
+ *
+ * @since X.X.X
+ */
+class PostTypeSniff extends AbstractFunctionParameterSniff {
+
+	const MAX_POST_TYPE_LENGTH = 20;
+
+	/**
+	 * Array of functions that must be checked.
+	 *
+	 * @since X.X.X
+	 *
+	 * @var array Multidimensional array with parameter details.
+	 *     $target_functions = array(
+	 *         (string) Function name.
+	 *     );
+	 */
+	protected $target_functions = array(
+		'register_post_type'   => true,
+	);
+
+	/**
+	 * Array of reserved post type names which can not be used by themes and plugins.
+	 *
+	 * @since X.X.X
+	 *
+	 * @var array
+	 */
+	protected $reserved_names = array(
+		'post'                => true,
+		'page'                => true,
+		'attachment'          => true,
+		'revision'            => true,
+		'nav_menu_item'       => true,
+		'custom_css'          => true,
+		'customize_changeset' => true,
+		'oembed_cache'        => true,
+		'user_request'        => true,
+		'wp_block'            => true,
+		'action'              => true,
+		'author'              => true,
+		'order'               => true,
+		'theme'               => true,
+	);
+
+	/**
+	 * Process the parameter of a matched function.
+	 *
+	 * Errors on invalid post type names when reserved keywords are used,
+	 * the post type is too long, or contains invalid characters.
+	 *
+	 * @since X.X.X
+	 *
+	 * @param int    $stackPtr        The position of the current token in the stack.
+	 * @param array  $group_name      The name of the group which was matched.
+	 * @param string $matched_content The token content (function name) which was matched.
+	 * @param array  $parameters      Array with information about the parameters.
+	 *
+	 * @return void
+	 */
+	public function process_parameters( $stackPtr, $group_name, $matched_content, $parameters ) {
+		$post_type = $this->strip_quotes( $parameters[1]['raw'] );
+		if ( isset( $this->reserved_names[ $post_type ] ) ) {
+			$this->phpcsFile->addError(
+				'%s() called with reserved post type %s. Reserved post types should not be used as they interfere with other WordPress functions.',
+				$stackPtr,
+				$this->string_to_errorcode( $post_type . '_Reserved' ),
+				array(
+					$matched_content,
+					$parameters[1]['raw'],
+				)
+			);
+		}
+
+		if ( strlen( $post_type ) > self::MAX_POST_TYPE_LENGTH ) {
+			$this->phpcsFile->addError(
+				'%s() called with invalid post type %s. Post type longer than %d characters.',
+				$stackPtr,
+				$this->string_to_errorcode( $post_type . '_TooLong' ),
+				array(
+					$matched_content,
+					$parameters[1]['raw'],
+					self::MAX_POST_TYPE_LENGTH,
+				)
+			);
+		}
+
+		if ( preg_match( '/[A-Z _]/', $post_type ) === 1 ) {
+			$this->phpcsFile->addError(
+				'%s() called with invalid post type %s. Post type contains invalid characters.',
```

Comment:
1. No need to use variable replacement for the function name.
2. Might it be an idea to be explicit about what characters are allowed in the error message (as per the above discussion) ?